blockchain/blockchain.go

```diff
@@ -608,82 +667,225 @@ func (b *Blockchain) WriteHeadersWithBodies(headers []*types.Header) error {
-		return fmt.Errorf(
+		b.logger.Error(fmt.Sprintf(
 			"transaction root hash mismatch: have %s, want %s",
 			hash,
 			block.Header.TxRoot,
-		)
+		))
+
+		return ErrInvalidTxRoot
 	}
 
-	// Checks are passed, write the chain
+	// Execute the transactions in the block and grab the result
+	blockResult, executeErr := b.executeBlockTransactions(block)
+	if executeErr != nil {
+		return fmt.Errorf("unable to execute block transactions, %w", executeErr)
+	}
+
+	// Verify the local execution result with the proposed block data
+	if err := blockResult.verifyBlockResult(&referenceBlockResult{
+		stateRoot:       block.Header.StateRoot,
+		receiptsRoot:    block.Header.ReceiptsRoot,
+		numTransactions: len(block.Transactions),
+		gasUsed:         block.Header.GasUsed,
+	}); err != nil {
+		return fmt.Errorf("unable to verify block execution result, %w", err)
+	}
+
+	return nil
+}
+
+// verifyBlockResult verifies that the block transaction execution result
+// matches up to the expected values
+func (br *BlockResult) verifyBlockResult(result *referenceBlockResult) error {
+	// Make sure the number of receipts matches the number of transactions
+	if len(br.Receipts) != result.numTransactions {
+		return ErrInvalidReceiptsSize
+	}
+
+	// Make sure the world state root matches up
+	if br.Root != result.stateRoot {
+		return ErrInvalidStateRoot
+	}
+
+	// Make sure the gas used is valid
+	if br.TotalGas != result.gasUsed {
+		return ErrInvalidGasUsed
+	}
+
+	// Make sure the receipts root matches up
+	receiptsRoot := buildroot.CalculateReceiptsRoot(br.Receipts)
+	if receiptsRoot != result.receiptsRoot {
+		return ErrInvalidReceiptsRoot
+	}
+
+	return nil
+}
+
+// executeBlockTransactions executes the transactions in the block locally,
+// and reports back the block execution result
+func (b *Blockchain) executeBlockTransactions(block *types.Block) (*BlockResult, error) {
 	header := block.Header
 
-	// Process and validate the block
-	res, err := b.processBlock(block)
+	parent, ok := b.readHeader(header.ParentHash)
+	if !ok {
+		return nil, ErrParentNotFound
+	}
+
+	blockCreator, err := b.consensus.GetBlockCreator(header)
 	if err != nil {
-		return err
+		return nil, err
+	}
+
+	txn, err := b.executor.ProcessBlock(parent.StateRoot, block, blockCreator)
+	if err != nil {
+		return nil, err
 	}
 
+	if err := b.consensus.PreStateCommit(header, txn); err != nil {
+		return nil, err
+	}
+
+	_, root := txn.Commit()
+
+	// Append the receipts to the receipts cache
+	b.receiptsCache.Add(header.Hash, txn.Receipts())
+
+	return &BlockResult{
+		Root:     root,
+		Receipts: txn.Receipts(),
+		TotalGas: txn.TotalGas(),
+	}, nil
+}
+
+// WriteBlock writes a single block
+func (b *Blockchain) WriteBlock(block *types.Block) error {
+	// Log the information
+	b.logger.Info(
+		"write block",
+		"num",
+		block.Number(),
+		"parent",
+		block.ParentHash(),
+	)
+
+	// Checks are passed, write the chain
```

Comment:
This comment seems redundant now.

Reply:
Removed 👍 

[a7c2300](https://github.com/0xPolygon/polygon-edge/pull/544/commits/a7c23001b14dda29acbba26ac50f327326cc54dd)